examples/structured_data/feature_space_advanced.py

```diff
@@ -0,0 +1,526 @@
+skew](https://developers.google.com/machine-learning/guides/rules-of-ml#training-serving_skew).
+skew](https://developers.google.com/machine-learning/guides/rules-of-ml#training-serving_skew).
+"""
+
+feature_space.save("myfeaturespace.keras")
+
+"""
+## Preprocessing with `FeatureSpace` as part of the tf.data pipeline
+
+We will opt to use our component asynchronously by making it part of the tf.data
+pipeline, as noted at the [previous
+guide](https://keras.io/examples/structured_data/structured_data_classification_with_featu
+re_space/) This enables asynchronous parallel preprocessing of the data on CPU before it
+hits the model. Usually, this is always the right thing to do during training.
+
+Let's create a training and validation dataset of preprocessed batches:
+"""
+
+preprocessed_train_ds = train_ds.map(
+    lambda x, y: (feature_space(x), y), num_parallel_calls=tf.data.AUTOTUNE
+).prefetch(tf.data.AUTOTUNE)
+
+preprocessed_valid_ds = valid_ds.map(
+    lambda x, y: (feature_space(x), y), num_parallel_calls=tf.data.AUTOTUNE
+).prefetch(tf.data.AUTOTUNE)
+
+"""
+## Model
+
+We will take advantage of our `FeatureSpace` component to build the model, as we want the
+model to be compatible with our preprocessing function, let's use the the `FeatureSpace`
+feature map as the input of our model.
+"""
+
+encoded_features = feature_space.get_encoded_features()
+print(encoded_features)
+
+"""
+This model is quite trivial only for demonstration purposes so don't pay too much
+attention to the architecture.
+"""
+
+x = tf.keras.layers.Dense(32, activation="relu")(encoded_features)
+x = tf.keras.layers.Dropout(0.5)(x)
+output = tf.keras.layers.Dense(1, activation="sigmoid")(x)
+
+model = tf.keras.Model(inputs=encoded_features, outputs=output)
+model.compile(optimizer="adam", loss="binary_crossentropy", metrics=["accuracy"])
+
+"""
+## Training
+
+Let's train our model for 50 epochs. Note that feature preprocessing is happening as part
+of the tf.data pipeline, not as part of the model.
+"""
+
+model.fit(
+    preprocessed_train_ds, validation_data=preprocessed_valid_ds, epochs=20, verbose=2
+)
+
+"""
+## Inference on new data with the end-to-end model
+
+Now, we can build our inference model (which includes the `FeatureSpace`) to make
+predictions based on dicts of raw features values, as follows:
+"""
+
+"""
+### Loading the `FeatureSpace`
+
+First let's load the `FeatureSpace` that we saved a few moment ago, this can be quite
+handy if you train a model but want to do inference at different time, possibly using a
+different device or environment.
+"""
+
+loaded_feature_space = tf.keras.models.load_model("myfeaturespace.keras")
+
+"""
+### Building the inference end-to-end model
+
+To build the inference model we need both the feature input map and the preprocessing
+encoded Keras tensors.
+"""
+
+dict_inputs = loaded_feature_space.get_inputs()
+encoded_features = loaded_feature_space.get_encoded_features()
+print(encoded_features)
+
+dict_inputs
```

Comment:
Always use print statements to display values since the file should work as a standalone Python script